{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wczytanie i przygotowanie danych"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pliki `.bed`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "negative_set = pd.read_csv('negative_set.bed', delimiter='\\t', header=None, names=[\"Chromosome\", \"StartPos\", \"EndPos\"])\n",
    "positive_set = pd.read_csv('GM12878.bed', delimiter='\\t', header=None, names=[\"Chromosome\", \"StartPos\", \"EndPos\", \"Score\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>StartPos</th>\n",
       "      <th>EndPos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>32251</td>\n",
       "      <td>36771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>39431</td>\n",
       "      <td>39891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>72312</td>\n",
       "      <td>74222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>84717</td>\n",
       "      <td>85177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>90499</td>\n",
       "      <td>91949</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  StartPos  EndPos\n",
       "0      chr10     32251   36771\n",
       "1      chr10     39431   39891\n",
       "2      chr10     72312   74222\n",
       "3      chr10     84717   85177\n",
       "4      chr10     90499   91949"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negative_set.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>StartPos</th>\n",
       "      <th>EndPos</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>773300</td>\n",
       "      <td>774100</td>\n",
       "      <td>7.866088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778980</td>\n",
       "      <td>779450</td>\n",
       "      <td>6.472419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>800100</td>\n",
       "      <td>802000</td>\n",
       "      <td>11.010675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>825670</td>\n",
       "      <td>826410</td>\n",
       "      <td>6.114487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>839470</td>\n",
       "      <td>842590</td>\n",
       "      <td>8.848865</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  StartPos  EndPos      Score\n",
       "0       chr1    773300  774100   7.866088\n",
       "1       chr1    778980  779450   6.472419\n",
       "2       chr1    800100  802000  11.010675\n",
       "3       chr1    825670  826410   6.114487\n",
       "4       chr1    839470  842590   8.848865"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positive_set.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dodanie etykiety do zbiorów\n",
    "negative_set[\"IsEnhancer\"] = 0\n",
    "positive_set[\"IsEnhancer\"] = 1\n",
    "\n",
    "positive_set = positive_set.drop(columns=\"Score\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# połączenie zbiorów danych\n",
    "total_set = pd.concat([negative_set, positive_set])\n",
    "\n",
    "# usunięcie nieprawidłowych chromosomów\n",
    "total_set = total_set.loc[total_set[\"Chromosome\"].str.match(r\"chr\\d+\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>StartPos</th>\n",
       "      <th>EndPos</th>\n",
       "      <th>IsEnhancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>32251</td>\n",
       "      <td>36771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>39431</td>\n",
       "      <td>39891</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>72312</td>\n",
       "      <td>74222</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>84717</td>\n",
       "      <td>85177</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>90499</td>\n",
       "      <td>91949</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48555</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140651150</td>\n",
       "      <td>140652330</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48556</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140702130</td>\n",
       "      <td>140703100</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48557</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140703310</td>\n",
       "      <td>140704120</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48558</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140710520</td>\n",
       "      <td>140711890</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48559</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140719390</td>\n",
       "      <td>140720520</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127039 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome   StartPos     EndPos  IsEnhancer\n",
       "0          chr10      32251      36771           0\n",
       "1          chr10      39431      39891           0\n",
       "2          chr10      72312      74222           0\n",
       "3          chr10      84717      85177           0\n",
       "4          chr10      90499      91949           0\n",
       "...          ...        ...        ...         ...\n",
       "48555       chr9  140651150  140652330           1\n",
       "48556       chr9  140702130  140703100           1\n",
       "48557       chr9  140703310  140704120           1\n",
       "48558       chr9  140710520  140711890           1\n",
       "48559       chr9  140719390  140720520           1\n",
       "\n",
       "[127039 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_set"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plik `.fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# odczyt pliku za pomocą biblioteki biopython\n",
    "\n",
    "fasta_sequences = {}\n",
    "\n",
    "with open(\"GRCh37.primary_assembly.genome.fa\", \"r\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        fasta_sequences[record.id] = record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr1', 'chr10', 'chr11', 'chr12', 'chr13']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mamy podział na chromosomy\n",
    "list(fasta_sequences.keys())[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_sequences[\"chr1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# na podstawie danych z plików .bed\n",
    "# wybieramy odpowiednie fragmenty DNA\n",
    "\n",
    "dna_sequence_list = []\n",
    "\n",
    "for index, data in total_set.iterrows():\n",
    "    chromosome = data[\"Chromosome\"]\n",
    "    start_pos = data[\"StartPos\"]\n",
    "    end_pos = data[\"EndPos\"]\n",
    "    dna_sequence = str(fasta_sequences[chromosome][start_pos:end_pos])\n",
    "    dna_sequence_list.append(dna_sequence)\n",
    "\n",
    "total_set[\"DNA sequence\"] = dna_sequence_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>StartPos</th>\n",
       "      <th>EndPos</th>\n",
       "      <th>IsEnhancer</th>\n",
       "      <th>DNA sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>32251</td>\n",
       "      <td>36771</td>\n",
       "      <td>0</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>39431</td>\n",
       "      <td>39891</td>\n",
       "      <td>0</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>72312</td>\n",
       "      <td>74222</td>\n",
       "      <td>0</td>\n",
       "      <td>CTGCCACCGTGCCTGGCTAATTTCTGTATTTTTAGTAGAGAGGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>84717</td>\n",
       "      <td>85177</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAAGTTGAATCTCTGAATAGACCAATAACAGGATCTGAAATTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>90499</td>\n",
       "      <td>91949</td>\n",
       "      <td>0</td>\n",
       "      <td>AGGTCAGGTGATCTGCAGCCAACCAAGCAGCTGCTAAGTGGCCAAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  StartPos  EndPos  IsEnhancer  \\\n",
       "0      chr10     32251   36771           0   \n",
       "1      chr10     39431   39891           0   \n",
       "2      chr10     72312   74222           0   \n",
       "3      chr10     84717   85177           0   \n",
       "4      chr10     90499   91949           0   \n",
       "\n",
       "                                        DNA sequence  \n",
       "0  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...  \n",
       "1  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...  \n",
       "2  CTGCCACCGTGCCTGGCTAATTTCTGTATTTTTAGTAGAGAGGGGG...  \n",
       "3  AGAAGTTGAATCTCTGAATAGACCAATAACAGGATCTGAAATTGTG...  \n",
       "4  AGGTCAGGTGATCTGCAGCCAACCAAGCAGCTGCTAAGTGGCCAAC...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_set.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# usuwanie rekordów z 'N' w sekwencji DNA\n",
    "total_set = total_set[~total_set[\"DNA sequence\"].str.contains(\"N\")]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### k-mery"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>StartPos</th>\n",
       "      <th>EndPos</th>\n",
       "      <th>IsEnhancer</th>\n",
       "      <th>DNA sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>72312</td>\n",
       "      <td>74222</td>\n",
       "      <td>0</td>\n",
       "      <td>CTGCCACCGTGCCTGGCTAATTTCTGTATTTTTAGTAGAGAGGGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>84717</td>\n",
       "      <td>85177</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAAGTTGAATCTCTGAATAGACCAATAACAGGATCTGAAATTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>90499</td>\n",
       "      <td>91949</td>\n",
       "      <td>0</td>\n",
       "      <td>AGGTCAGGTGATCTGCAGCCAACCAAGCAGCTGCTAAGTGGCCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr10</td>\n",
       "      <td>104967</td>\n",
       "      <td>107977</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTTCCCCACTTGGTATGATACTAGCTGTGGGTCTGTCATATATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr10</td>\n",
       "      <td>114029</td>\n",
       "      <td>114779</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGAGCTGTTAAAGCACTATGTGTGGGTATGTCTGTGAAGGTGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48555</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140651150</td>\n",
       "      <td>140652330</td>\n",
       "      <td>1</td>\n",
       "      <td>TTGGGGTAGAGATTTCATCTTTGTCATCTTGGCTGTAATGTGCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48556</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140702130</td>\n",
       "      <td>140703100</td>\n",
       "      <td>1</td>\n",
       "      <td>TGTATTTTTAGTAGAGGCGGGGTTTCACCACGTTGCCCAGGCTGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48557</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140703310</td>\n",
       "      <td>140704120</td>\n",
       "      <td>1</td>\n",
       "      <td>GTTGCAGTGAACCAAGATCGTGCCAACTGCATTCCAGCCTGGGTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48558</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140710520</td>\n",
       "      <td>140711890</td>\n",
       "      <td>1</td>\n",
       "      <td>GGCAGCTTCCTGCCGGAGCCCCACATTCTGCTCGTATTAGCACGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48559</th>\n",
       "      <td>chr9</td>\n",
       "      <td>140719390</td>\n",
       "      <td>140720520</td>\n",
       "      <td>1</td>\n",
       "      <td>CCGTGCAGCTTTGTTTTCTGGAAGGTCGGGCATAATGCTTGTTCTC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120892 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome   StartPos     EndPos  IsEnhancer  \\\n",
       "2          chr10      72312      74222           0   \n",
       "3          chr10      84717      85177           0   \n",
       "4          chr10      90499      91949           0   \n",
       "5          chr10     104967     107977           0   \n",
       "6          chr10     114029     114779           0   \n",
       "...          ...        ...        ...         ...   \n",
       "48555       chr9  140651150  140652330           1   \n",
       "48556       chr9  140702130  140703100           1   \n",
       "48557       chr9  140703310  140704120           1   \n",
       "48558       chr9  140710520  140711890           1   \n",
       "48559       chr9  140719390  140720520           1   \n",
       "\n",
       "                                            DNA sequence  \n",
       "2      CTGCCACCGTGCCTGGCTAATTTCTGTATTTTTAGTAGAGAGGGGG...  \n",
       "3      AGAAGTTGAATCTCTGAATAGACCAATAACAGGATCTGAAATTGTG...  \n",
       "4      AGGTCAGGTGATCTGCAGCCAACCAAGCAGCTGCTAAGTGGCCAAC...  \n",
       "5      TTTTCCCCACTTGGTATGATACTAGCTGTGGGTCTGTCATATATGA...  \n",
       "6      GAGAGCTGTTAAAGCACTATGTGTGGGTATGTCTGTGAAGGTGTTT...  \n",
       "...                                                  ...  \n",
       "48555  TTGGGGTAGAGATTTCATCTTTGTCATCTTGGCTGTAATGTGCAAA...  \n",
       "48556  TGTATTTTTAGTAGAGGCGGGGTTTCACCACGTTGCCCAGGCTGGT...  \n",
       "48557  GTTGCAGTGAACCAAGATCGTGCCAACTGCATTCCAGCCTGGGTGA...  \n",
       "48558  GGCAGCTTCCTGCCGGAGCCCCACATTCTGCTCGTATTAGCACGTA...  \n",
       "48559  CCGTGCAGCTTTGTTTTCTGGAAGGTCGGGCATAATGCTTGTTCTC...  \n",
       "\n",
       "[120892 rows x 5 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tutaj trzeba policzyć k-mery\n",
    "# i sconcatować kolumny do tej ramki\n",
    "\n",
    "total_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Kmers_cols(column, k):\n",
    "    kmers = itertools.product('ACTG', repeat=k)\n",
    "    df = pd.DataFrame()\n",
    "    for i in kmers:\n",
    "        kmer = ''.join(i)\n",
    "        rev_kmer = kmer.replace('A', '%temp%').replace('T', 'A').replace('%temp%', 'T').replace('C', '%temp%').replace('G', 'C').replace('%temp%', 'G')\n",
    "        if(rev_kmer not in df.columns):\n",
    "            df[kmer] = column.str.count(kmer)/(len(column.iloc[0])-k+1)\n",
    "        else:\n",
    "            df[rev_kmer] += column.str.count(kmer)/(len(column.iloc[0])-k+1)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAAA</th>\n",
       "      <th>AAAC</th>\n",
       "      <th>AAAT</th>\n",
       "      <th>AAAG</th>\n",
       "      <th>AACA</th>\n",
       "      <th>AACC</th>\n",
       "      <th>AACT</th>\n",
       "      <th>AACG</th>\n",
       "      <th>AATA</th>\n",
       "      <th>AATC</th>\n",
       "      <th>...</th>\n",
       "      <th>CGCT</th>\n",
       "      <th>CGCG</th>\n",
       "      <th>CGTA</th>\n",
       "      <th>CGTC</th>\n",
       "      <th>CGTT</th>\n",
       "      <th>CGTG</th>\n",
       "      <th>CGGA</th>\n",
       "      <th>CGGC</th>\n",
       "      <th>CGGT</th>\n",
       "      <th>CGGG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.018353</td>\n",
       "      <td>0.012061</td>\n",
       "      <td>0.019927</td>\n",
       "      <td>0.012061</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>0.008915</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.013110</td>\n",
       "      <td>0.006293</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.004719</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.008915</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.006817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.005768</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.004719</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>0.004719</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.005768</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.006293</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>0.005768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.042475</td>\n",
       "      <td>0.027268</td>\n",
       "      <td>0.039853</td>\n",
       "      <td>0.026744</td>\n",
       "      <td>0.024122</td>\n",
       "      <td>0.014158</td>\n",
       "      <td>0.015732</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>0.026744</td>\n",
       "      <td>0.009439</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.006293</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.007341</td>\n",
       "      <td>0.006293</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.004719</td>\n",
       "      <td>0.001573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.005768</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.002098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>0.019402</td>\n",
       "      <td>0.009439</td>\n",
       "      <td>0.015732</td>\n",
       "      <td>0.010488</td>\n",
       "      <td>0.007866</td>\n",
       "      <td>0.005768</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>0.006293</td>\n",
       "      <td>0.008390</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.005768</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.002098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>0.005768</td>\n",
       "      <td>0.004719</td>\n",
       "      <td>0.009439</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.009439</td>\n",
       "      <td>0.006293</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.000524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.003671</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.005768</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.003146</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.001049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>0.017829</td>\n",
       "      <td>0.016780</td>\n",
       "      <td>0.024646</td>\n",
       "      <td>0.010488</td>\n",
       "      <td>0.011012</td>\n",
       "      <td>0.011536</td>\n",
       "      <td>0.008915</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.017305</td>\n",
       "      <td>0.011012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001573</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006293</td>\n",
       "      <td>0.011012</td>\n",
       "      <td>0.005768</td>\n",
       "      <td>0.005244</td>\n",
       "      <td>0.002098</td>\n",
       "      <td>0.004195</td>\n",
       "      <td>0.007341</td>\n",
       "      <td>0.005768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 128 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         AAAA      AAAC      AAAT      AAAG      AACA      AACC      AACT  \\\n",
       "2    0.018353  0.012061  0.019927  0.012061  0.008390  0.008390  0.008915   \n",
       "3    0.005768  0.001573  0.004195  0.003146  0.002098  0.002098  0.002622   \n",
       "4    0.002098  0.003671  0.001049  0.005244  0.004719  0.006817  0.005244   \n",
       "5    0.042475  0.027268  0.039853  0.026744  0.024122  0.014158  0.015732   \n",
       "6    0.002098  0.002098  0.003671  0.004195  0.001573  0.002098  0.002098   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "97   0.019402  0.009439  0.015732  0.010488  0.007866  0.005768  0.005244   \n",
       "98   0.002098  0.000524  0.003671  0.001573  0.003146  0.001573  0.000524   \n",
       "99   0.005768  0.004719  0.009439  0.004195  0.006817  0.003671  0.002622   \n",
       "100  0.004195  0.002098  0.003671  0.002622  0.003146  0.000524  0.000000   \n",
       "101  0.017829  0.016780  0.024646  0.010488  0.011012  0.011536  0.008915   \n",
       "\n",
       "         AACG      AATA      AATC  ...      CGCT      CGCG      CGTA  \\\n",
       "2    0.003671  0.013110  0.006293  ...  0.004195  0.002098  0.003671   \n",
       "3    0.000000  0.006817  0.004719  ...  0.000000  0.000000  0.001049   \n",
       "4    0.003146  0.002098  0.001573  ...  0.001573  0.000000  0.003146   \n",
       "5    0.008390  0.026744  0.009439  ...  0.000524  0.000524  0.006293   \n",
       "6    0.001049  0.005768  0.002622  ...  0.000000  0.000000  0.002098   \n",
       "..        ...       ...       ...  ...       ...       ...       ...   \n",
       "97   0.006293  0.008390  0.005244  ...  0.000524  0.001573  0.003146   \n",
       "98   0.001049  0.000524  0.002098  ...  0.001049  0.000000  0.000524   \n",
       "99   0.003146  0.009439  0.006293  ...  0.000000  0.000000  0.003671   \n",
       "100  0.000524  0.005768  0.002622  ...  0.000000  0.000524  0.000000   \n",
       "101  0.006817  0.017305  0.011012  ...  0.001573  0.000000  0.006293   \n",
       "\n",
       "         CGTC      CGTT      CGTG      CGGA      CGGC      CGGT      CGGG  \n",
       "2    0.004719  0.006817  0.006817  0.008915  0.004195  0.006817  0.006817  \n",
       "3    0.001049  0.003146  0.000524  0.000524  0.000524  0.000524  0.000000  \n",
       "4    0.005768  0.002622  0.004195  0.006293  0.001049  0.005244  0.005768  \n",
       "5    0.004195  0.007341  0.006293  0.003671  0.001049  0.004719  0.001573  \n",
       "6    0.001573  0.002622  0.001049  0.001573  0.001573  0.000524  0.002098  \n",
       "..        ...       ...       ...       ...       ...       ...       ...  \n",
       "97   0.005768  0.002622  0.004195  0.003671  0.000000  0.002098  0.002098  \n",
       "98   0.000524  0.002098  0.003671  0.000000  0.002098  0.000524  0.000524  \n",
       "99   0.003146  0.001049  0.001573  0.002098  0.000000  0.001049  0.000524  \n",
       "100  0.001049  0.001049  0.002622  0.003146  0.001049  0.002622  0.001049  \n",
       "101  0.011012  0.005768  0.005244  0.002098  0.004195  0.007341  0.005768  \n",
       "\n",
       "[100 rows x 128 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# liczy się do dosyć wolno\n",
    "\n",
    "import warnings\n",
    "\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\")\n",
    "    kmers = Kmers_cols(total_set[\"DNA sequence\"].head(100), k=4)\n",
    "    \n",
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# podział na zbiory treningowy i testowy\n",
    "\n",
    "test_set = total_set.loc[total_set[\"Chromosome\"].isin([\"chr1\", \"chr14\", \"chr21\"])]\n",
    "trening_set = total_set.loc[~total_set[\"Chromosome\"].isin([\"chr1\", \"chr14\", \"chr21\"])]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py310",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
